File: python/leetcode/pancake_sorting.py
class Solution(object):
    def pancakeSort(self, A):
        """
        :type A: List[int]
        :rtype: List[int]
        """
        res = []
        for i in range(len(A)-1, -1, -1):
            k = 0
            while A[k] != i+1: k += 1
            if k == i: continue
            if k != 0:
                self.flip(A, k+1)
                res.append(k+1)
            self.flip(A, i+1)
            res.append(i+1)
        return res

    def flip(self, A, k):
        for i in range(k/2):
            A[i], A[k-1-i] = A[k-1-i], A[i]

File: src/main/java/LeetCode/PancakeSorting.java
package LeetCode;

public class PancakeSorting {
    public List<Integer> pancakeSort(int[] A) {
        List<Integer> res = new ArrayList();
        int n = A.length;
        for (int i = n-1; i >= 0; i--) {
            int k = 0;
            while (A[k] != i+1) k++;
            if (k == i) continue;
            if (k != 0) {
                flip(A, k+1);
                res.add(k+1);
            }
            flip(A, i+1);
            res.add(i+1);
        }
        return res;
    }

    private void flip(int[] A, int k) {
        for (int i = 0; i < k/2; i++) {
            int tmp = A[i];
            A[i] = A[k-1-i];
            A[k-1-i] = tmp;
        }
    }
}
